### PR TITLE
Reschedule Feature #193

### PR DESCRIPTION
**Pull request by albertdatui**
I have implemented the reschedule feature together with its test cases.

The reschedule function supports
* rescheduling from different types of spec (e.g: date to object, Reccurence Rule to date, etc)
* rescheduling canceled job, job that has been executed (date), etc.

In addition, the method does not emit 'rescheduled' event since I am trying to minimize the modification to the code.

I also changed schedule-cron-jobs.js (I found some bugs there), just to pass all of the testcase.

### CODE REVIEW

File: src/schedule.js [hunk 1]
```diff
@@ -446,6 +463,9 @@ function scheduleNextRecurrence(rule, job, prevDate) {
   return inv;
 }
 
+/* Convenience methods */
+var scheduledJobs = {};
+
```
Comment (santigimeno) [inline src/schedule.js]: why this change?

File: src/schedule.js [hunk 2]
```diff
@@ -498,6 +515,20 @@ function scheduleJob() {
...
 
+function rescheduleJob(job, spec) {
+  var success = false;
+  if (job instanceof Job) {
+    success = job.reschedule(spec);
+  } else {
+    var findJob = getJob(job);
+    if (findJob instanceof Job) {
+      return rescheduleJob(findJob, spec);
+    }
```
Comment (santigimeno) [inline src/schedule.js]: success = rescheduleJob...

File: src/schedule.js [hunk 3]
```diff
@@ -111,6 +111,23 @@ function Job() {
...
+        inv = cInvs[j];
+
+        cancelInvocation(inv);
+        this.stopTrackingInvocation(inv);
+      }
+
+      return true;
+    }
+
+    return null;
```
Comment (santigimeno) [inline src/schedule.js]: return false?

File: src/schedule.js (hunk 2 above)
Comment (albertdatui) [inline src/schedule.js]: I have no objection for this one, I think the output will be the same.

File: src/schedule.js (hunk 3 above)
Comment (albertdatui) [inline src/schedule.js]: This is purely my mistake, it should be false instead of null.

File: src/schedule.js (hunk 1 above)
Comment (albertdatui) [inline src/schedule.js]: for this one, I just want to pass the npm run lint.

File: src/schedule.js [hunk 4]
```diff
@@ -498,6 +515,20 @@ function scheduleJob() {
...
+  } else {
+    var findJob = getJob(job);
+    if (findJob instanceof Job) {
+      return rescheduleJob(findJob, spec);
+    }
+  }
+
+  return success;
+}
+
```
Comment (santigimeno) [inline src/schedule.js]: How about this function returns the `Job` instance in the same way as `scheduleJob` does?

File: test/convenience-method-test.js [hunk 5]
```diff
@@ -212,6 +212,362 @@ module.exports = {
...
+
+      var rule = new schedule.RecurrenceRule();
+      rule.second = null; // fire every second
+
+      var job = schedule.scheduleJob(rule, function() {
+        test.ok(true);
+      });
+
+      setTimeout(function() {
+        schedule.rescheduleJob("Blah", new Date(Date.now() + 2000));
```
Comment (santigimeno) [inline test/convenience-method-test.js]: check that returns null?

File: test/convenience-method-test.js [hunk 6]
```diff
@@ -212,6 +212,362 @@ module.exports = {
...
+          minute: null
+        });
+      }, 3250);
+
+      setTimeout(function() {
+        job.cancel();
+        test.done();
+      }, 5000);
+
+      clock.tick(5000);
```
Comment (santigimeno) [inline test/convenience-method-test.js]: All these tests are great. It would be a nice addition to be able to check the # of times the job triggers before and after the rescheduling is performed

**Comment by santigimeno** [comment]
Apart from the comments it LGTM

**Comment by santigimeno** [comment]
Thanks BTW

File: test/convenience-method-test.js (hunk 6 above)
Comment (albertdatui) [inline test/convenience-method-test.js]: I think it is a good idea. However, do you have any idea on how to do it?
One way that I can think of is to add one more field into Job object to count the number of invocation.

Any other suggestion?

**Comment by albertdatui** [comment]
@santigimeno, the previous bug (cron-parser) keep failing me to pass all of the test.
Will it be better to address that bug first and then continue the enhancement?

File: test/convenience-method-test.js (hunk 6 above)
Comment (santigimeno) [inline test/convenience-method-test.js]: I think adding a field it's a good idea

**Comment by santigimeno** [comment]
Yes, I think so

**Comment by tejasmanohar** [comment]
@albertdatui Other bug is fixed now. Feel free to change the code based on @santigimeno's comments, and then, we'll merge :D - thanks so much man!

**Comment by albertdatui** [comment]
@tejasmanohar @santigimeno Thanks!! I will try to merge it as soon as possible. I have been pretty busy these days.

 I will give you a heads up. :)

**Comment by santigimeno** [comment]
Duplicate of https://github.com/node-schedule/node-schedule/pull/258 that was already merged
